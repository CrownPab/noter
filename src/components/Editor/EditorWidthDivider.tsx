import { Box, Divider } from "@chakra-ui/react";
import { Dispatch, SetStateAction } from "react";

interface EditorWidthDividerProps {
  setPaddingX: Dispatch<SetStateAction<number>>;
  setIsBeingDragged: Dispatch<SetStateAction<boolean>>;
  isBeingDragged?: boolean;
}

export default ({
  setPaddingX,
  setIsBeingDragged,
  isBeingDragged = false,
}: EditorWidthDividerProps) => {
  return (
    <Box position="relative">
      <Divider
        _before={
          isBeingDragged
            ? {
                content: `""`,
                position: "absolute",
                top: 0,
                left: "-3px",
                bottom: 0,
                border: "3px solid",
              }
            : {}
        }
        borderColor="blackAlpha.800"
        position="absolute"
        top="0"
        bottom="0"
        px="5px"
        orientation="vertical"
        draggable="true"
        onDragStart={(e) => {
          // This removes the `ghost` image that appears while dragging
          e.dataTransfer.setDragImage(new Image(), 0, 0);
          setIsBeingDragged(true);
        }}
        onDrag={(e) => {
          if (e.clientX > 200 && e.clientX < window.innerWidth - 200) {
            let paddingX = e.clientX;
            setPaddingX(paddingX);
          }
        }}
        onDragEnd={() => {
          setIsBeingDragged(false);
        }}
        cursor="ew-resize"
      />
    </Box>
  );
};
