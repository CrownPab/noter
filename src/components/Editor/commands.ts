import {
  type EditorState,
  type Transaction,
  type TransactionSpec,
} from '@codemirror/state'

export const modifyLines = (
  state: EditorState,
  dispatch: (tr: Transaction) => void,
  textModifier: (text: string) => string
): TransactionSpec => {
  const { from, to } = state.selection.main
  const lineFrom = state.doc.lineAt(from)
  const lineTo = state.doc.lineAt(to)
  const text = state.doc.sliceString(lineFrom.from, lineTo.to)

  const textToInsert = textModifier(text)

  dispatch(
    state.update({
      changes: {
        from: lineFrom.from,
        to: lineTo.to,
        insert: textToInsert,
      },
      selection: {
        anchor: lineFrom.from + textToInsert.length,
        head: lineFrom.from + textToInsert.length,
      },
    })
  )

  return {}
}

export const modifySelection = (
  state: EditorState,
  dispatch: (tr: Transaction) => void,
  textModifier: (text: string) => string
): TransactionSpec => {
  const text: string = state.doc
    .slice(state.selection.main.from, state.selection.main.to)
    .toString()

  const textToInsert = textModifier(text)

  dispatch(
    state.update({
      changes: {
        from: state.selection.main.from,
        to: state.selection.main.to,
        insert: textToInsert,
      },
    })
  )
  return {}
}

export const toggleBold = (text: string): string => {
  if (text.startsWith('**') && text.endsWith('**')) {
    return text.slice(2, -2)
  }
  return `**${text}**`
}

export const toggleItalic = (text: string): string => {
  if (text.startsWith('_') && text.endsWith('_')) {
    return text.slice(1, -1)
  }
  return `_${text}_`
}

export const toggleStrikeThrough = (text: string): string => {
  if (text.startsWith('~~') && text.endsWith('~~')) {
    return text.slice(2, -2)
  }
  return `~~${text}~~`
}

const toggleCheckbox = (text: string): string => {
  if (text.startsWith('- [x] ')) {
    return text.replace('- [x] ', '- [ ] ')
  }
  if (text.startsWith('- [ ] ')) {
    return text.replace('- [ ] ', '- [x] ')
  }
  return `- [ ] ${text}`
}

export const toggleCheckList = (text: string): string => {
  return text.split('\n').map(toggleCheckbox).join('\n')
}

export const toggleCheckListAll = (text: string): string => {
  return toggleCheckbox(text)
}
