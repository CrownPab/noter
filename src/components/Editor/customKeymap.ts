import { Prec } from '@codemirror/state'
import { keymap } from '@codemirror/view'
import {
  modifySelection,
  toggleBold,
  toggleItalic,
} from '~/components/Editor/commands'

export default Prec.highest(
  keymap.of([
    {
      key: 'Ctrl-i',
      run: (editor) => {
        modifySelection(editor.state, editor.dispatch, toggleItalic)
        return true
      },
      preventDefault: true,
    },
    {
      key: 'Ctrl-b',
      run: (editor) => {
        modifySelection(editor.state, editor.dispatch, toggleBold)
        return true
      },
    },
  ])
)
