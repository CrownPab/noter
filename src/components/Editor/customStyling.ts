import { HighlightStyle, syntaxHighlighting } from '@codemirror/language'
import { tags } from '@lezer/highlight'

export default syntaxHighlighting(
  HighlightStyle.define([
    { tag: tags.heading1, fontSize: '155%' },
    { tag: tags.heading2, fontSize: '145%' },
    { tag: tags.heading3, fontSize: '135%' },
    { tag: tags.heading4, fontSize: '125%' },
    { tag: tags.heading5, fontSize: '115%' },
    { tag: tags.heading6, fontSize: '105%' },
  ])
)
