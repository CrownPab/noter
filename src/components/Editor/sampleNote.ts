export const sampleNote = `
# Heading level 1
## Heading level 2
### Heading level 3
#### Heading level 4
##### Heading level 5
###### Heading level 6


Just ~~some~~ _**random**_ **_text_** here.

Normal test here for testing. Lorem Ipsum 

- [ ] Task 1
- [x] Task 2
`
