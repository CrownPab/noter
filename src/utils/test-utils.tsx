import { cleanup, render, type RenderResult } from '@testing-library/react'
import { afterEach } from 'vitest'

afterEach(() => {
  cleanup()
})

const customRender = (ui: JSX.Element, options = {}): RenderResult =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => children,
    ...options,
  })

// eslint-disable-next-line import/export
export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// eslint-disable-next-line import/export
export { customRender as render }
