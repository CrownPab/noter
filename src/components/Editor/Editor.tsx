import { Box } from '@chakra-ui/react'
import { markdown, markdownLanguage } from '@codemirror/lang-markdown'
import { languages } from '@codemirror/language-data'

import { defaultHighlightStyle, syntaxHighlighting } from '@codemirror/language'
import React, { useEffect, useRef } from 'react'
import customKeymap from '~/components/Editor/customKeymap'
import customStyling from '~/components/Editor/customStyling'

import { basicSetup, EditorView } from 'codemirror'
import { EditorState } from '@codemirror/state'
import { sampleNote } from '~/components/Editor/sampleNote'

interface EditorProps {
  _onInit?: (_: EditorView) => void
  _onUpdate?: () => void
}

const Editor = ({
  _onInit = (_: EditorView) => {},
  _onUpdate = () => {},
}: EditorProps): JSX.Element => {
  const editorRef = useRef<HTMLDivElement>(null)

  useEffect(() => {
    const view = new EditorView({
      state: EditorState.create({
        doc: sampleNote,
        extensions: [
          basicSetup,
          markdown({
            base: markdownLanguage,
            codeLanguages: languages,
          }),
          syntaxHighlighting(defaultHighlightStyle),
          customStyling,
          customKeymap,
          EditorView.theme({
            '&': {
              height: '100%',
            },
          }),
        ],
      }),
      parent: editorRef.current as HTMLDivElement,
    })
    if (editorRef.current !== null) {
      _onInit(view)
    }
    return () => {
      view.destroy()
    }
  }, [])

  return <Box w="100%" h="100%" ref={editorRef} data-testid="editor"></Box>
}

export default Editor
