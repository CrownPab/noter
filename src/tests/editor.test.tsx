import { beforeEach, describe, it } from 'vitest'
import { render } from '~/utils/test-utils'
import Editor from '~/components/Editor/Editor'
import { type EditorView } from 'codemirror'
import { sampleNote } from '~/components/Editor/sampleNote'

describe('Test the Editor', () => {
  let editor: EditorView

  beforeEach(() => {
    render(
      <Editor
        _onInit={(_editor) => {
          editor = _editor
        }}
      />
    )
  })

  it('should have the sample note', ({ expect }) => {
    expect(editor.state.doc.toString()).toBe(sampleNote)
  })

  it('should render text correctly', ({ expect }) => {
    expect(editor.contentDOM).toMatchSnapshot()
  })
})
