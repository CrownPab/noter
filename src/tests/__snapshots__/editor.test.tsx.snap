// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test the Editor > should render text correctly 1`] = `
<div
  aria-autocomplete="list"
  aria-multiline="true"
  autocapitalize="off"
  autocorrect="off"
  class="cm-content"
  contenteditable="true"
  data-language="markdown"
  role="textbox"
  spellcheck="false"
  style="tab-size: 4"
  translate="no"
>
  <div
    class="cm-activeLine cm-line"
  >
    <br />
  </div>
  <div
    class="cm-line"
  >
    <span
      class="ͼ7 ͼo ͼ5"
    >
      #
    </span>
    <span
      class="ͼ7 ͼo"
    >
       Heading level 1
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="ͼ7 ͼp ͼ5"
    >
      ##
    </span>
    <span
      class="ͼ7 ͼp"
    >
       Heading level 2
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="ͼ7 ͼq ͼ5"
    >
      ###
    </span>
    <span
      class="ͼ7 ͼq"
    >
       Heading level 3
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="ͼ7 ͼr ͼ5"
    >
      ####
    </span>
    <span
      class="ͼ7 ͼr"
    >
       Heading level 4
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="ͼ7 ͼs ͼ5"
    >
      #####
    </span>
    <span
      class="ͼ7 ͼs"
    >
       Heading level 5
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="ͼ7 ͼt ͼ5"
    >
      ######
    </span>
    <span
      class="ͼ7 ͼt"
    >
       Heading level 6
    </span>
  </div>
  <div
    class="cm-line"
  >
    <br />
  </div>
  <div
    class="cm-line"
  >
    <br />
  </div>
  <div
    class="cm-line"
  >
    Just 
    <span
      class="ͼa ͼ5"
    >
      ~~
    </span>
    <span
      class="ͼa"
    >
      some
    </span>
    <span
      class="ͼa ͼ5"
    >
      ~~
    </span>
     
    <span
      class="ͼ8 ͼ5"
    >
      _
    </span>
    <span
      class="ͼ8 ͼ9 ͼ5"
    >
      **
    </span>
    <span
      class="ͼ8 ͼ9"
    >
      random
    </span>
    <span
      class="ͼ8 ͼ9 ͼ5"
    >
      **
    </span>
    <span
      class="ͼ8 ͼ5"
    >
      _
    </span>
     
    <span
      class="ͼ9 ͼ5"
    >
      **
    </span>
    <span
      class="ͼ9 ͼ8 ͼ5"
    >
      _
    </span>
    <span
      class="ͼ9 ͼ8"
    >
      text
    </span>
    <span
      class="ͼ9 ͼ8 ͼ5"
    >
      _
    </span>
    <span
      class="ͼ9 ͼ5"
    >
      **
    </span>
     here.
  </div>
  <div
    class="cm-line"
  >
    <br />
  </div>
  <div
    class="cm-line"
  >
    Normal test here for testing. Lorem Ipsum 
  </div>
  <div
    class="cm-line"
  >
    <br />
  </div>
  <div
    class="cm-line"
  >
    <span
      class="ͼ5"
    >
      -
    </span>
     
    <span
      class="ͼc"
    >
      [ ]
    </span>
     Task 1
  </div>
  <div
    class="cm-line"
  >
    <span
      class="ͼ5"
    >
      -
    </span>
     
    <span
      class="ͼc"
    >
      [x]
    </span>
     Task 2
  </div>
  <div
    class="cm-line"
  >
    <br />
  </div>
</div>
`;
